ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/BackupReplayFile.java

```diff
@@ -79,16 +89,39 @@ public String getPath() {
     return file.getAbsolutePath();
   }
 
-  public void write(final ConsumerRecord<byte[], byte[]> record) throws IOException {
-    if (writer == null) {
-      throw new IOException("Write permission denied.");
-    }
-
+  private static void appendRecordToFile(
+      final ConsumerRecord<byte[], byte[]> record,
+      final File file,
+      final Filesystem filesystem
+  ) throws IOException {
+    final FileOutputStream writer = createWriter(file, filesystem);
     writer.write(record.key());
     writer.write(KEY_VALUE_SEPARATOR_BYTES);
     writer.write(record.value());
     writer.write(NEW_LINE_SEPARATOR_BYTES);
-    writer.flush();
+    writer.close();
+  }
+
+  public void write(final ConsumerRecord<byte[], byte[]> record) throws IOException {
+    if (!writable) {
+      throw new IOException("Write permission denied.");
+    }
+    final File dirty = dirty(file);
+    final File tmp = tmp(file);
+    // first write to the dirty copy
+    appendRecordToFile(record, dirty, filesystem);
+    // atomically rename the dirty copy to the "live" copy while copying the live copy to
+    // the "dirty" copy via a temporary hard link
+    Files.createLink(tmp.toPath(), file.toPath());
+    Files.move(
+        dirty.toPath(),
+        file.toPath(),
+        StandardCopyOption.REPLACE_EXISTING,
+        StandardCopyOption.ATOMIC_MOVE
+    );
+    Files.move(tmp.toPath(), dirty.toPath());
+    // keep the dirty copy in sync with the live copy, which now has the write
+    appendRecordToFile(record, dirty, filesystem);
```

Comment:
and this is the new dirty copy, right? Which was previously the live copy? It's a bit confusing since L124 and L 112 are the exact same